slides/E-async-web.md

```diff
@@ -0,0 +1,892 @@
+
+---
+layout: default
+---
+
+# The `Future` is now!
+
+```rust
+pub trait Future {
+    type Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
+}
+```
+
+*Question: what stands out to you?*
+<v-click>
+<div>
+
+- `&mut self` &rarr; `Pin<&mut Self>`: makes `Self` immovable
+- `wake: fn()` &rarr; `cx: &mut Context<'_>`: contains a `Waker`
+
+*More on `Pin<&mut Self>` in the [Rust async book](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html)*
+</div>
+</v-click>
+
+---
+layout: section
+---
+
+# `async` and `await`
+
+---
+layout: default
+---
+
+# Expanding `async`
+&nbsp;  
+
+**" `Futures` are cool, but why didn't I see them in the web scraper example?"*
+
+`async fn`s and `async` blocks are syntactic sugar generating `Future`
+<v-click>
+<div>
+
+```rust
+async fn foo() -> u8 { 5 }
+```
+</div>
+</v-click>
+<v-click>
+<div>
+
+is equivalent to:
+```rust
+fn foo() -> impl Future<Output=u8> {
+    async {
+        5
+    }
+}
+```
+</div>
+</v-click>
+<v-click>
+<div>
+
+which is equivalent to:
+```rust
+fn foo() -> impl Future<Output=u8> {
+    /// Create a future that is immediately ready with a value.
+    futures::future::ready(5)
+}
+```
+</div>
+</v-click>
+
+---
+layout: default
+---
+
+# Expanding `async` and `await`
+
+```rust
+let fut_one = /* ... */;
+let fut_two = /* ... */;
+async move {
+    fut_one.await;
+    fut_two.await;
+}
+```
+<v-click>
+<div>
+Generates an opaque type implementing `Future`:
+
+```rust
+struct AsyncFuture {
+    fut_one: FutOne,
+    fut_two: FutTwo,
+    state: State,
+}
+enum State {
+    AwaitingFutOne,
+    AwaitingFutTwo,
+    Done,
+}
+```
+
+*This and the following is not the actually generated code, but it's a good mental model*
+</div>
+</v-click>
+
+
+---
+layout: default
+---
+
+# Expanding `async` and `await` (2)
+
+```rust
+impl Future for AsyncFuture {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        loop {
+            match self.state {
+                State::AwaitingFutOne => match self.fut_one.poll(/* - snip - */) {
+                    Poll::Ready(()) => self.state = State::AwaitingFutTwo,
+                    Poll::Pending => return Poll::Pending,
+                }
+                State::AwaitingFutTwo => match self.fut_two.poll(/* - snip - */) {
+                    Poll::Ready(()) => self.state = State::Done,
+                    Poll::Pending => return Poll::Pending,
+                }
+                State::Done => return Poll::Ready(()),
+            }
+        }
+    }
+}   
+```
+
+Kind of looks like `AndThenFut`!
+
+*Adapted from [Asynchronous programming in Rust](https://rust-lang.github.io/async-book/03_async_await/01_chapter.html)*
+
+---
+layout: default
+---
+
+# `async`/`await` expansion takeaways
+
+- Rust generates state machines out of `await`s
+- Generated code may become very complex, but original is easy to follow
```

Comment:
You could add some more drama/context here; students have seen this trick before with the `?` syntax: it would be annoying to write the Error-handling boilerplate it hides over and over again, but you *could do it*.

In the async case *it is theoretically possible* to do it, but you really *wouldn't want to do it* because you would get horrible programs with lots of bugs and you would go mad.